src/org/etools/j1939_84/bus/RP1210Bus.java

```diff
@@ -323,42 +285,45 @@ private Packet decode(byte[] data, int length) {
         return buf;
     }
 
-    private BusLogger getLogger() {
-        return logger;
-    }
-
     /**
      * Checks the {@link RP1210Library} for any incoming messages. Any incoming
      * messages are decoded and added to the queue
      */
     private void poll() {
         try {
             while (true) {
+                byte[] data = new byte[32];
                 short rtn = rp1210Library.RP1210_ReadMessage(clientId, data, (short) data.length, BLOCKING_NONE);
                 if (rtn > 0) {
-                    Packet packet = decode(data, rtn);
-                    logger.log(Level.FINE, packet::toTimeString, null);
-                    if (packet.getSource() == getAddress() && !packet.isTransmitted()) {
-                        logger.log(Level.WARNING, () -> "Another ECU is using this address: " + packet, null);
-                        imposterDetected = true;
-                    }
-                    queue.add(packet);
+                    decodeDataAndQueuePacket(data, rtn);
                 } else if (rtn == -RP1210Library.ERR_RX_QUEUE_FULL) {
                     // RX queue full, remedy is to reread.
-                    logger.log(Level.SEVERE, () -> getErrorMessage(rtn), null);
+                    logger.log(Level.SEVERE, getErrorMessage(rtn));
                 } else {
                     checkReturnCode(rtn);
                     break;
                 }
             }
             // this allows the other calls to have a chance
             Thread.yield();
```

Comment:
I didn’t notice this before, but I’m not sure the `yield()` is really doing any good.  When the next runnable is submitted, the thread has already yielded. Right?

Reply:
yield() tells the thread management (which is probably the OS) that now would be a good time to switch threads.  It's only a hint.  It is independent of the task list that executor is using to queue up tasks.  Without this, the executor has no reason to pause between submitted tasks, which without any sends would effectively be a tight loop on poll().  Any context switch then would be less predictably after an empty read.  This is only a suggestion to context switch from this thread; it may be anytime.